src/refiners/foundationals/fuyu/fuyu.py

```diff
@@ -0,0 +1,267 @@
+        network = create_fuyu(config)
+        ```
+    """
+    model = Fuyu(
+        embedding_dim=config.embedding_dim,
+        feedforward_dim=config.feedforward_dim,
+        max_sequence_length=config.max_sequence_length,
+        vocabulary_size=config.vocabulary_size,
+        tokenizer=config.tokenizer,
+        patch_size=config.patch_size,
+        padding_value=config.padding_value,
+        num_layers=config.num_layers,
+        num_heads=config.num_heads,
+        norm_eps=config.norm_eps,
+        base=config.base,
+        partial_rotary_factor=config.partial_rotary_factor,
+        use_bias=config.use_bias,
+        is_optimized=config.is_optimized,
+        device=config.device,
+        dtype=config.dtype,
+    )
+    return model
+
+
+class Fuyu(fl.Chain):
+    """
+    Implements the Fuyu model
+    see [https://www.adept.ai/blog/fuyu-8b]
+    """
+
+    def __init__(
+        self,
+        embedding_dim: int,
+        feedforward_dim: int,
+        max_sequence_length: int,
+        vocabulary_size: int,
+        tokenizer: FuyuTokenizer,
+        patch_size: int,
+        padding_value: float,
+        num_layers: int,
+        num_heads: int,
+        norm_eps: float,
+        base: int,
+        partial_rotary_factor: float,
+        use_bias: bool,
+        is_optimized: bool,
+        device: Device | str | None,
+        dtype: DType | None,
+    ) -> None:
+        self.tokenizer = [tokenizer]
+        self.input_encoder = [
+            InputEncoder(
+                embedding_dim=embedding_dim,
+                max_sequence_length=max_sequence_length,
+                vocabulary_size=vocabulary_size,
+                tokenizer=self.tokenizer[0],
+                patch_size=patch_size,
+                padding_value=padding_value,
+                device=device,
+                dtype=dtype,
+            )
+        ]
+
+        super().__init__(
+            self.input_encoder[0],
+            FuyuTransformer(
+                FuyuTransformerLayer(
+                    embedding_dim=embedding_dim,
+                    feedforward_dim=feedforward_dim,
+                    num_heads=num_heads,
+                    norm_eps=norm_eps,
+                    base=base,
+                    partial_rotary_factor=partial_rotary_factor,
+                    use_bias=use_bias,
+                    is_optimized=is_optimized,
+                    device=device,
+                    dtype=dtype,
+                )
+                for _ in range(num_layers)
+            ),
+            fl.LayerNorm(normalized_shape=embedding_dim, eps=norm_eps, device=device, dtype=dtype),
+            fl.Linear(in_features=embedding_dim, out_features=vocabulary_size, bias=False, device=device, dtype=dtype),
+        )
+
+    def init_context(self) -> dict[str, dict[str, Any]]:
+        return {"attention": {"mask": None}}
+
+    def generate(self, images: List[Image.Image], prompts: List[str], max_len_generation: int = 50):
```

Comment:
nit, please add a unit test, and maybe break down the function a bit

Reply:
unit test added + function broken down